### PR TITLE
Add support for Authenticated Users

### PR DESCRIPTION
**Pull request by lukabratos**
### What?
Offer a secure method to directly target a notification to an authenticated user in your application.

### Why?
This can be useful if you want to send personalized notifications, for example, if you have an events application, you could send invites targeting specific users knowing that these notifications can’t be accessed by other users.

### CODE REVIEW

File: .rubocop.yml [hunk 1]
```diff
@@ -5,19 +5,19 @@ AllCops:
   TargetRubyVersion: 2.4.0
 
 Metrics/LineLength:
-  Max: 100
+  Enabled: false
```
Comment (luismfonseca) [inline .rubocop.yml:12]: Why do we need lines longer than this?

File: README.md [hunk 2]
```diff
@@ -52,7 +54,17 @@ data = {
   }
 }
 
-Pusher::PushNotifications.publish(interests: ['hello'], payload: data)
+# Publish the given 'data' to the specified interests.
+Pusher::PushNotifications.publish_to_interests(interests: ['hello'], payload: data)
+
+# Publish the given 'data' to the specified users.
+Pusher::PushNotifications.publish_to_users(users: ['jonathan', 'jordan', 'luis', 'luka', 'mina'], payload: data)
```
Comment (luismfonseca) [inline README.md:61]: no `cucas`? :(

File: README.md [hunk 3]
```diff
@@ -66,7 +78,7 @@ The errors statuses can be:
 | HTTP Status | Reason                                                                                        |
 | ----------- | --------------------------------------------------------------------------------------------- |
 | 401         | Incorrect secret key                                                                          |
-| 400         | Payload too big (10Kb limit), Payload invalid, Payload in a wrong schema, instance_id missing |
+| 422         | Payload too big (10Kb limit), Payload invalid, Payload in a wrong schema, instance_id missing |
```
Comment (luismfonseca) [inline README.md]: We also have a 429 Too Many Requests.

Maybe we should point to the complete (and always up-to-date) list: https://docs.pusher.com/beams/reference/publish-api#error-responses

File: lib/pusher/push_notifications/client.rb [hunk 4]
```diff
@@ -40,15 +55,20 @@ def post(resource, payload = {})
...
       def_delegators :@config, :instance_id, :secret_key
 
-      def build_url(resource)
-        "https://#{instance_id}.#{BASE_URL}/#{instance_id}/#{resource}"
+      def build_publish_url(resource)
+        "https://#{instance_id}.pushnotifications.pusher.com/publish_api/v1/instances/#{instance_id}/#{resource}"
+      end
+
+      def build_users_url(resource)
+        "https://#{instance_id}.pushnotifications.pusher.com/user_api/v1/instances/#{instance_id}/#{resource}"
```
Comment (luismfonseca) [inline lib/pusher/push_notifications/client.rb]: It's no longer `user_api` it's `customer_api`.

File: lib/pusher/push_notifications/client.rb [hunk 5]
```diff
@@ -40,15 +55,20 @@ def post(resource, payload = {})
...
       end
 
       def headers
         {
           content_type: 'application/json',
           accept: :json,
-          Authorization: "Bearer #{secret_key}"
+          Authorization: "Bearer #{secret_key}",
+          'X-Pusher-Library': "push-notifications-server-ruby #{Pusher::PushNotifications::VERSION}"
```
Comment (luismfonseca) [inline lib/pusher/push_notifications/client.rb]: `pusher-push-notifications-ruby <version>`

File: lib/pusher/push_notifications/client.rb [hunk 6]
```diff
@@ -30,7 +28,24 @@ def post(resource, payload = {})
             body = JSON.parse(response.body)
             Response.new(status, body, status == 200)
           else
-            Response.new(500, nil, false)
+            Response.new(status, nil, false)
+          end
+        end
+      end
+
+      def delete(resource)
```
Comment (luismfonseca) [inline lib/pusher/push_notifications/client.rb]: It's odd to call it `resource` when you already know that 100% of the cases it's a `user`.

File: lib/pusher/push_notifications/client.rb [hunk 7]
```diff
@@ -30,7 +28,24 @@ def post(resource, payload = {})
...
+        url = build_users_url(resource)
```
Comment (luismfonseca) [inline lib/pusher/push_notifications/client.rb]: We need to URL encode the user id as we support any UTF-8 characters (e.g. we allow the user id `/`).

Please add a test for this as well.

File: lib/pusher/push_notifications/user_id.rb [hunk 8]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+module Pusher
+  module PushNotifications
+    class UserId
+      def max_user_id_length
```
Comment (luismfonseca) [inline lib/pusher/push_notifications/user_id.rb]: It's a bit odd to have these as functions as they will never change their value

File: spec/pusher/push_notifications/generate_token_spec.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+          Pusher::PushNotifications::UseCases::GenerateToken::GenerateTokenError
+        ).with_message(
+          'User id length too long (expected fewer than 165 characters)'
+        )
+      end
+    end
+
+    context 'when user id is valid' do
+      it 'will generate the token' do
+        exp = Time.now.to_i + 24 * 60 * 60 # Current time + 24h
```
Comment (luismfonseca) [inline spec/pusher/push_notifications/generate_token_spec.rb]: I think this test might be flaky no? as it implies that this line and the line where it checks for the token runs in the same second.

I think I'd just check that the generated token is a valid JWT and that the `sub` is the user id.

File: spec/pusher/push_notifications/generate_token_spec.rb [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        expect { generate_token }.to raise_error(
+          Pusher::PushNotifications::UseCases::GenerateToken::GenerateTokenError
+        ).with_message(
+          'User Id cannot be empty.'
+        )
+      end
+    end
+
+    context 'when user id is too long' do
+      let(:user) { 'a' * 165 }
```
Comment (luismfonseca) [inline spec/pusher/push_notifications/generate_token_spec.rb]: You can grab the length from the constant and then do a + 1. That means if you want to change the size, you'd only ever need to change 1 place and the tests would still pass.

File: spec/pusher/push_notifications/publish_to_users_spec.rb [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+        VCR.use_cassette('publishes/users/valid_users') do
+          response = publish_to_users
+
+          expect(response).to be_ok
+        end
+      end
+    end
+
+    context 'when too many user ids are provided' do
+      int_array = (1..1001).to_a.shuffle
```
Comment (luismfonseca) [inline spec/pusher/push_notifications/publish_to_users_spec.rb]: I think these values should be fetched from the constants and not hard-coded in the tests.

File: .rubocop.yml (hunk 1 above)
Comment (lukabratos) [reply .rubocop.yml:12]: I've set the max line length to 80 characters. Fixed in 64d765b.

File: README.md (hunk 3 above)
Comment (lukabratos) [reply README.md]: Done in 7d2b898.

File: lib/pusher/push_notifications/client.rb (hunk 4 above)
Comment (lukabratos) [reply lib/pusher/push_notifications/client.rb]: Fixed in 14bddc4.

File: lib/pusher/push_notifications/client.rb (hunk 5 above)
Comment (lukabratos) [reply lib/pusher/push_notifications/client.rb]: Fixed in 00bfb3c.

File: lib/pusher/push_notifications/client.rb (hunk 6 above)
Comment (lukabratos) [reply lib/pusher/push_notifications/client.rb]: Fixed in 0177cc5.

File: lib/pusher/push_notifications/user_id.rb (hunk 8 above)
Comment (lukabratos) [reply lib/pusher/push_notifications/user_id.rb]: Fixed in 339db5d.

File: spec/pusher/push_notifications/generate_token_spec.rb (hunk 10 above)
Comment (lukabratos) [reply spec/pusher/push_notifications/generate_token_spec.rb]: Fixed in 2d514d9.

File: spec/pusher/push_notifications/publish_to_users_spec.rb (hunk 11 above)
Comment (lukabratos) [reply spec/pusher/push_notifications/publish_to_users_spec.rb]: Fixed in [2d514d9](https://github.com/pusher/push-notifications-ruby/commit/2d514d94eb362d6c95a7b3951f436a35d71b3264).

File: spec/pusher/push_notifications/generate_token_spec.rb (hunk 9 above)
Comment (lukabratos) [reply spec/pusher/push_notifications/generate_token_spec.rb]: Fixed in 9316a09.

File: lib/pusher/push_notifications/client.rb (hunk 7 above)
Comment (lukabratos) [reply lib/pusher/push_notifications/client.rb]: Fixed in 259e79c.

**Comment by lukabratos** [comment]
This feature was released in version `1.1.0`.

**Pull request merged**